examples/permissioned-token/test/tests.js

```diff
@@ -0,0 +1,334 @@
+
+    // Refresh Accounts
+    syncInfo();
+  });
+
+  it('Token Issuance', () => {
+    // Cannot issue before opting-in
+    assert.throws(() => issue(elon, 20), 'RUNTIME_ERR1404');
+
+    // Can issue after optin-in
+    optInToASA(elon);
+    // syncInfo();
+    issue(elon, 20);
+    assert.equal(20, runtime.getAssetHolding(assetIndex, elon.address).amount);
+
+    // Cannot issue after killing token
+    killToken();
+    assert.throws(() => issue(elon, 20), 'RUNTIME_ERR1009');
+  });
+
+  it('WhiteListing', () => {
+    // Only asset manager can whitelist
+    const address = elon.address;
+    optInToPermissions(address);
+    assert.isDefined(elon.getAppFromLocal(permissionsAppId));
+
+    assert.throws(() => runtime.executeTx({
+      type: types.TransactionType.CallNoOpSSC,
+      sign: types.SignType.SecretKey,
+      fromAccount: bob.account, // Bob is not the asset manager
+      appId: permissionsAppId,
+      payFlags: { totalFee: 1000 },
+      appArgs: ['str:add_whitelist', `int:${controllerAppID}`],
+      accounts: [address],
+      foreignAssets: [assetIndex],
+      foreignApps: [controllerAppID]
+    }), 'RUNTIME_ERR1009');
+
+    whitelist(address);
+    assert.equal(
+      Number(elon.getLocalState(permissionsAppId, 'whitelisted')),
+      1
+    );
+  });
+
+  it('Token Transfer', () => {
+    fund(elon.address);
+    fund(bob.address);
+    fund(alice.address);
+    fund(clawbackAddress);
+    const amount = 20;
+
+    // Issue some tokens to Bob and Elon
+    optInToASA(elon);
+    optInToASA(bob);
+    issue(elon, 50);
+    issue(bob, 50);
+
+    // Cannot transfer directly
+    assert.throws(() => runtime.executeTx({
+      type: types.TransactionType.TransferAsset,
+      sign: types.SignType.SecretKey,
+      fromAccount: bob.account,
+      toAccountAddr: elon.address,
+      amount: amount,
+      assetID: assetIndex,
+      payFlags: {}
+    }), 'RUNTIME_ERR1505');
+
+    // Cannot transfer before being whitelisted
+    assert.throws(() => transfer(bob, elon, amount), 'RUNTIME_ERR1009');
+
+    // Can transfer after being whitelisted
+    whitelist(elon.address);
+    whitelist(bob.address);
+    assert.isDefined(elon.getAppFromLocal(permissionsAppId));
+    assert.isDefined(bob.getAppFromLocal(permissionsAppId));
+    assert.equal(
+      Number(elon.getLocalState(permissionsAppId, 'whitelisted')),
+      1
+    );
+    assert.equal(
+      Number(bob.getLocalState(permissionsAppId, 'whitelisted')),
+      1
+    );
+    const elonBalance = runtime.getAssetHolding(assetIndex, elon.address).amount;
+    const bobBalance = runtime.getAssetHolding(assetIndex, elon.address).amount;
+    transfer(bob, elon, amount);
+    assert.equal(
+      Number(runtime.getAssetHolding(assetIndex, elon.address).amount),
+      Number(elonBalance) + amount
+    );
+    assert.equal(
+      Number(runtime.getAssetHolding(assetIndex, elon.address).amount),
+      Number(bobBalance) + amount
+    );
+  });
+});
```

Comment:
Please add more tests and divide them in different files if needed, so it could be easily understandable for a developer

Reply:
Okay